### PR TITLE
Enhance verification code for the `tritongpu.extract` operator

### LINKED ISSUE #1623: [FlashAttention]: tritongpu.extract operation missing verification when result rank != operand rank
Add verification code to diagnose invalid use of the operation when:
  - the result rank is smaller than the operand rank
  - the result rank is greater than the result rank but the input tensor cannot be evenly partitioned by the result tensor shape.

### PR DESCRIPTION
**Pull request by etiotto**
Fixes #1623.

### CODE REVIEW

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td [hunk 1]
```diff
@@ -71,17 +71,25 @@ def TTIG_GlueOp : TTIG_Op<"glue", [Pure]> {
...
-    element type as the input tensor.
+    The `extract` operation extracts a subtensor (or ptr to subtensor) from an input tensor
+    (or ptr to tensor) as specified by the given index.
+    The value extracted has shape as specified by the result type and the same element type
+    as the input tensor. The operand must have the same element type as the result and rank
+    that is equal to or greater than the result rank.
 
     Example:
      ```mlir
+      // Partition tensor<32x32xf16>into 4 16x16xf16 subtensors and extract the
```
Comment (whitneywhtsang) [inline third_party/intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td:82]:
```suggestion
      // Partition tensor<32x32xf16> into 4 16x16xf16 subtensors and extract the
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Ops.cpp [hunk 2]
```diff
@@ -145,16 +145,20 @@ LogicalResult ExtractOp::verify() {
...
+    return emitOpError("result rank cannot be greater than operand rank");
 
-  // ensure the input can be partitioned by the requested result.
   SmallVector<int64_t> resultShape = getShape(resultType);
   SmallVector<int64_t> operandShape = getShape(operandType);
 
+  // Make the result have the same rank as the operand.
+  while (resultRank < operandRank) {
+    resultShape.insert(resultShape.begin(), operandRank - resultRank, 1);
+    resultRank++;
```
Comment (whitneywhtsang) [inline third_party/intel/lib/Dialect/TritonIntelGPU/IR/Ops.cpp:157]:
```suggestion
    ++resultRank;
```

**Review by whitneywhtsang — APPROVED** [review]

**Comment by Dewei-Wang-sh** [comment]
```
f32, f32... f32 <=> 16xf32
16xf32, 16xf32, ..., 16xf32  <=> 8x16xf32
```
as stated in match-target-size pass, the operands should not be constrained to tensor, it can be scalar.

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td (hunk 1 above)
Comment (etiotto) [reply third_party/intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td:82]: done

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Ops.cpp (hunk 2 above)
Comment (etiotto) [reply third_party/intel/lib/Dialect/TritonIntelGPU/IR/Ops.cpp:157]: no difference

**Comment by etiotto** [comment]
> ```
> f32, f32... f32 <=> 16xf32
> 16xf32, 16xf32, ..., 16xf32  <=> 8x16xf32
> ```
> 
> as stated in match-target-size pass, the operands should not be constrained to tensor, it can be scalar.

This PR doesn't constraint the input operand of the "extract" operation to be a tensor. If it is a tensor it does verify that the result shape requested can partition the input shape.

File: third_party/intel/backend/compiler.py [hunk 3]
```diff
@@ -49,6 +49,7 @@ class XPUOptions:
     max_num_imprecise_acc_default: int = 0  # `max_num_imprecise_acc` only applies to fp8 -> fp32 dot on sm_90 for cuda
     extern_libs: dict = None
     debug: bool = False
+    large_grf: bool = False
```
Comment (whitneywhtsang) [inline third_party/intel/backend/compiler.py:52]: Please remove this
```suggestion
```

File: third_party/intel/lib/Dialect/TritonIntelGPU/IR/Ops.cpp (hunk 2 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/Dialect/TritonIntelGPU/IR/Ops.cpp:157]: https://llvm.org/docs/CodingStandards.html#prefer-preincrement

File: third_party/intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td [hunk 4]
```diff
@@ -71,17 +71,25 @@ def TTIG_GlueOp : TTIG_Op<"glue", [Pure]> {
 def TTIG_ExtractOp : TTIG_Op<"extract", [Pure]> {
   let summary = "Tensor extract operation";
   let description = [{
-    The `extract` operation extracts a subtensor (or ptr to subtensor) from an
-    input tensor (or ptr to tensor) as specified by the given index.
-    The value extracted has shape as specified by the result type and the same
-    element type as the input tensor.
+    The `extract` operation extracts a subtensor (or ptr to subtensor) from an input tensor
```
Comment (Dewei-Wang-sh) [inline third_party/intel/include/Dialect/TritonIntelGPU/IR/TritonIntelGPUOps.td:74]: nit: subtensor or scalar,

**Pull request merged**